### ISSUE TITLE
Clip norm does not clip gradients to given value

### ISSUE THREAD

**Issue by shourya01** [APPFL/APPFL#151]
**Description:**
The gradient being normalized to a constant is important for implementation of differential privacy. In cases where there are multiple local epochs, the gradient may be interpreted as the difference of weights before and after local updates. However, enabling clip norm does not normalize it to the provided values.

**System specs**
Conda env with APPFL (built on commit 6237d40), PyTorch 2.0.1. System has NVIDIA Titan RTX GPU with cuda enabled.

**How to replicate:**
To `client_optimizer.py`, add after line 20 (i.e. `optimizer = eval(self.optim)(self.model.parameters(), **self.optim_args)`):
```
# ADDITIONAL
named_primal_states_old = torch.cat([copy.deepcopy(self.model.state_dict())[name].flatten() for name,_ in self.model.named_parameters()]).to(self.cfg.device)
```
Also add after line 70 (i.e. `self.round += 1`):
```
# ADDITIONAL
named_primal_states_new = torch.cat([copy.deepcopy(self.model.state_dict())[name].flatten() for name,_ in self.model.named_parameters()]).to(self.cfg.device)
# ADDITIONAL
print('%d-norm difference after weight_update on client %d is %.5f'%(self.clip_norm,self.id,torch.linalg.norm(named_primal_states_old-named_primal_states_new,ord=self.clip_norm).item()),flush=True)
```

Now run MNIST training as follows:
```
mpiexec -np 3 python /home/L2RPN/APPFL/APPFL/examples/mnist_mpi_sync.py --partition class_noiid --loss_fn losses/celoss.py --loss_fn_name CELoss --num_epochs 5 --clip_grad --clip_value 1 --local_train_pattern epochs
```
Replace `/home/L2RPN/APPFL/APPFL/examples` with path to your APPFL examples directory.

**Expected output:**
The `print` statement added above tells us that norm is 1.

**Observed output:**
```
-------Loading_Time= 25.968549966812134
-------Loading_Time= 26.06582760810852
-------Loading_Time= 26.132920026779175
1-norm difference after weight_update on client 1 is 10622.83789
1-norm difference after weight_update on client 0 is 11226.82617
      Iter   Local(s)  Global(s)   Valid(s) PerIter(s) Elapsed(s)   TestLoss   TestAccu
         1       5.07       0.00       2.25       7.33       7.33     1.0653      70.79
1-norm difference after weight_update on client 1 is 7932.06250
1-norm difference after weight_update on client 0 is 7710.22021
         2       2.50       0.00       0.24       2.75      10.08     0.6634      78.09
1-norm difference after weight_update on client 1 is 8705.10742
1-norm difference after weight_update on client 0 is 9565.17383
         3       2.63       0.00       0.24       2.88      12.96     0.6579      80.18
1-norm difference after weight_update on client 1 is 9470.45898
1-norm difference after weight_update on client 0 is 10335.47852
         4       2.60       0.00       0.25       2.85      15.81     0.6270      79.65
1-norm difference after weight_update on client 1 is 9529.82520
1-norm difference after weight_update on client 0 is 11480.05371
         5       2.55       0.00       0.26       2.82      18.63     0.6477      80.19
Device=cuda
#Processors=3
#Clients=2
Server=ServerFedAvg
Clients=ClientOptim
Comm_Rounds=5
Local_Rounds=1
DP_Eps=False
Clipping=1.0
Elapsed_time=18.63
BestAccuracy=80.19
client_learning_rate = 0.003 
------DONE------ 0
------DONE------ 1
------DONE------ 2
```

**Comment by Zilinghan** [comment]
@minseok-ryu Hi Minseok, could you please leave some suggestions here, thanks

**Comment by yim0331** [comment]
@shourya01 Is there any reason to use the CPU memory to calculate the torch.linalg.norm while GPU is available?

**Comment by shourya01** [comment]
@yim0331 I did it just as a workaround since the print statement is temporary.

**Comment by yim0331** [comment]
@shourya01 I believe it's better to operate the torch functions with GPU memory if the memory available.

**Comment by shourya01** [comment]
@yim0331 Edited to use gpu.

**Comment by Zilinghan** [comment]
I think the main concern here is that Shourya just wants to check whether the norm of the model weight difference is equal to clip_value (i.e., 1 in this case), and obviously, the code does not guarantee that as we are doing more than one epoch.

But as I am not very familiar with the DP theory, I am not sure how we should manipulate with the gradient (model difference). So I think we need @minseok-ryu  to help us confirm with this

**Comment by shourya01** [comment]
@minseok-ryu @Zilinghan Well said. Also want to highlight that this issue persists even if using `client_step_optimizer` and doing 1 `num_local_steps` and 1 `num_local_epochs` (i.e. just one backprop per server epoch locally). Upon adding the above code (to print norms) to `client_step_optimizer.py` and running the following:
```
mpiexec -np 3 python /home/L2RPN/APPFL/APPFL/examples/mnist_mpi_sync.py --partition class_noiid --loss_fn losses/celoss.py --loss_fn_name CELoss --num_epochs 5 --clip_grad --clip_value 1 --local_train_pattern steps --num_local_steps 1 --num_local_epochs 1
```
we still get the following result:
```
-------Loading_Time= 25.801041841506958
-------Loading_Time= 25.829277515411377
-------Loading_Time= 26.71730065345764
1-norm difference after weight_update on client 1 is 1261.56177
1-norm difference after weight_update on client 0 is 1245.69507
      Iter   Local(s)  Global(s)   Valid(s) PerIter(s) Elapsed(s)   TestLoss   TestAccu
         1       3.65       0.00       2.27       5.93       5.93     2.3952       9.74
1-norm difference after weight_update on client 1 is 1006.12640
1-norm difference after weight_update on client 0 is 1015.22363
         2       2.01       0.00       0.24       2.26       8.19     2.2693      28.57
1-norm difference after weight_update on client 0 is 855.14990
1-norm difference after weight_update on client 1 is 887.60925
         3       2.01       0.00       0.26       2.27      10.46     2.1907      18.71
1-norm difference after weight_update on client 1 is 998.03412
1-norm difference after weight_update on client 0 is 988.48004
         4       2.01       0.00       0.25       2.26      12.73     2.1008      20.77
1-norm difference after weight_update on client 1 is 1033.52783
1-norm difference after weight_update on client 0 is 1017.88239
         5       2.01       0.00       0.26       2.28      15.00     1.9904      33.04
Device=cuda
#Processors=3
#Clients=2
Server=ServerFedAvg
Clients=ClientStepOptim
Comm_Rounds=5
Local_Rounds=1
DP_Eps=False
Clipping=1.0
Elapsed_time=15.0
BestAccuracy=33.04
client_learning_rate = 0.003 
------DONE------ 0
------DONE------ 1
------DONE------ 2
```

**Comment by minseok-ryu** [comment]
If I remember it correctly, I clipped the model parameter, not the gradient.

**Comment by minseok-ryu** [comment]
The norm of the gradient is clipped by "1" in this case.

**Comment by shourya01** [comment]
@minseok-ryu Even after dividing by lr, the value is still like 300000 or so. Also you're right its different from norm of gradient in cases where the optimizer is not SGD (since adam, etc. maintain smoothing states). But ultimately I wanted to pin down that quantity which has norm 1 to which dp noise is added.

**Comment by minseok-ryu** [comment]
@shourya01 Have you checked the norm of gradient directly right after clipping to check if the function actually works?

**Comment by shourya01** [comment]
@minseok-ryu Sorry for the delay. I wrote a minimal working example to check for whether clipping works. Here is it, based on the latest release of APPFL:
```
import torch
import numpy as np
from mpi4py import MPI
from torch import nn as nn
import argparse
import logging
from appfl.misc.data import Dataset
from appfl.config import *
from appfl.misc.utils import set_seed
import appfl.run_mpi as rm

'''
In this script, we see whether setting clip_norm sets the norm of difference to
desired value.
'''

""" parser arguments """
parser = argparse.ArgumentParser()
parser.add_argument("--clip_update_norm", type=int, default=1, help="whether to clip or not")
parser.add_argument("--seed", type=int, default=1, help="seed for reproducability")
parser.add_argument("--batch_size", type=int, default=1, help="batch size for training")
parser.add_argument("--savedir", type=str, default='/home/L2RPN', help="dirname for saving model")
args = parser.parse_args()

def get_dataset(n_points,n_clients):
    
    """ return dataset corresponding to function y=3*x + 2 """
    
    all_x_points = np.linspace(0,1,n_points)
    split_x_points = np.array_split(all_x_points,n_clients)
    split_y_points = [3*itm+2 for itm in split_x_points]
    
    datasets = [Dataset(torch.FloatTensor(x.reshape(-1,1)),torch.FloatTensor(y.reshape(-1,1)))
                for x,y in zip(split_x_points,split_y_points)]
    retval = datasets if n_clients > 1 else datasets[0]
    
    return retval

def get_flattened_params(model):
    
    """ get numpy vector of learnable parameters of model """
    
    param_collector = []
    
    for _,val in model.named_parameters():
        param_collector.append(val.detach().cpu().numpy().reshape(-1))
        
    return np.concatenate(param_collector,axis=0)

if __name__ == "__main__":
    
    comm = MPI.COMM_WORLD
    comm_rank = comm.Get_rank()
    comm_size = comm.Get_size()
    
    assert comm_size == 3, "Hard coded 1 server and 2 clients for this example, comm_size is .%d"%comm_size
    
    model = nn.Sequential(nn.Linear(in_features=1,out_features=1),) # model
    loss_fn = nn.MSELoss() # loss
    metric = lambda y_true, y_pred: np.mean( np.abs(y_true-y_pred) ) # test metric
    num_clients = 2 # number of clients
    train_datasets = get_dataset(n_points = 1000, n_clients = num_clients) # get train dataset for clients
    test_datasets = None # disable test dataset
    
    cfg = OmegaConf.structured(Config)
    
    """ set cfg values to ensure experiment results are valid"""
    
    cfg.fed.args.clip_grad = True if args.clip_update_norm else False # clip grad according to input args
    cfg.fed.args.clip_value = 1 # set clip value
    cfg.fed.args.num_local_steps = 1 # set single local step
    cfg.fed.args.optim_args.lr = 1 # set client LR to 1
    cfg.fed.args.server_learning_rate = 1 # set server LR to 1
    
    cfg.train_data_batch_size = args.batch_size # set batch size according to input
    cfg.validation = False # set validation to be false
    
    cfg.fed.clientname = "ClientStepOptim" # use step-based optimizer
    cfg.num_epochs = 1 # set single global update
    cfg.save_model = True # save the model
    cfg.save_model_dirname = args.savedir # set save directory - CHANGE AS NEEDED
    
    
    set_seed(args.seed) # set random seed according to args for reproducability
    
    if comm_rank == 0:
        """ save model state before update """
        cur_param = get_flattened_params(model)
    
    if comm_rank == 0:
        """ run server """
        rm.run_server(cfg, comm, model, loss_fn, num_clients)
    else:
        """ run_client """
        rm.run_client(cfg, comm, model, loss_fn, num_clients, train_datasets[comm_rank - 1])
        
    if comm_rank == 0:
        model = torch.load(cfg.save_model_dirname+'/_Round_1.pt')
        new_param = get_flattened_params(model)
        print('old model params:')
        print(cur_param)
        print('new model params:')
        print(new_param)
        print('norm difference:')
        print(np.linalg.norm(cur_param-new_param,ord=1))
    ```
The result is as follows:
```
old model params:
[0.8536658  0.11722028]
new model params:
[ 0.6111026  -0.41244116]
norm difference:
0.77222466
```
I have set all step numbers, learning rates, and even batch size to 1. In this case, a DP implementation would require the norm to be equal to `clip_value` but it looks like here it is not true.

**Comment by shourya01** [comment]
Solved with #197
